heat/optim/dp_optimizer.py

```diff
@@ -1,6 +1,788 @@
+
+    @torch.no_grad()
+    def epoch_loss_logic(self, loss, loss_globally_averaged=False):
+        """
+        Function controlling the number of batches between global synchronizations and the batches to wait before
+        receiving the sent parameters. The warm-up and cool-down phases are also controlled here.
+
+        This function should be called at the end of each epoch with the training loss value at the end of the epoch.
+
+        The number of batches between local synchronizations can also be modified here with minor code adjustments.
+
+        Parameters
+        ----------
+        loss: torch.Tensor or float
+            loss value of the current epoch
+        loss_globally_averaged: bool, optional
+            boolean if the loss is already globally averaged
+        """
+
+        if not loss_globally_averaged:
+            loss_send = torch.zeros(self.comm.size)
+            # loss.data -> this will get the raw number from the lass value and nothing else
+            loss_send[self.comm.rank] = loss.data if isinstance(loss, torch.Tensor) else loss
+            self.comm.Allreduce(MPI.IN_PLACE, loss_send, MPI.SUM)
+            avg_loss = torch.mean(loss_send)
+        else:
+            avg_loss = torch.tensor(loss)
+
+        if self.epoch < self.warmup_epochs:
+            self.global_skip = 0
+            self.local_skip = 0
+            self.batches_to_wait = 0
+
+            self.print0(
+                f"Warmup Phase, parameters of next epoch\n\t Global Skips: {self.global_skip}, "
+                f" Local Skips {self.local_skip}, Batches to wait: {self.batches_to_wait}"
+            )
+            return
+
+        elif self.warmup_epochs == self.epoch:
+            self.global_skip = 4
+            self.local_skip = 1
+            self.batches_to_wait = 1
+
+            self.print0(
+                f"End of Warmup Phase, parameters of next epoch\n\t Global Skips: {self.global_skip}, "
+                f" Local Skips {self.local_skip}, Batches to wait: {self.batches_to_wait}"
+            )
+
+        if self.epoch >= self.total_epochs - self.cooldown_epochs:
+            self.global_skip = 0
+            self.local_skip = 0
+            self.batches_to_wait = 0
+
+            self.print0(
+                f"Cooldown Phase, parameters of next epoch:\n\tGlobal Skips: {self.global_skip}, "
+                f" Local Skips {self.local_skip:.4f}, Batches to wait: {self.batches_to_wait}"
+            )
+            return
+
+        if self.global_skip == self.max_gs and self.max_gs > 4:
+            self._gs8_waited += 1
+
+        self.print0(
+            f"Best loss value: {self.stability.best * (1.0 - self.stability.threshold):.4f}"
+            f" Current loss: {avg_loss:.4f}, Worse epochs: {self.stability.num_bad_epochs}"
+        )
+
+        stable = self.stability.test_if_improving(avg_loss)
+
+        if stable and self.global_skip > 1:
+            # drop gs by factor of 2
```

Comment:
im confused by this comment. what do you mean?

Reply:
The factor 2 should be a global drop constant at the top of the file or class constant so that it can be easily changed in the future. Same goes for the other constants below, like the // 4